### PR TITLE
red-knot: infer and display ellipsis type

### PR DESCRIPTION
**Pull request by chriskrycho**
## Summary

Just what it says on the tin: adds basic `EllipsisType` inference for any time `...` appears in the AST.

## Test Plan

Test that `x = ...` produces exactly what we would expect.

### CODE REVIEW

File: crates/red_knot_python_semantic/src/types.rs [hunk 1]
```diff
@@ -297,6 +299,10 @@ impl<'db> Type<'db> {
                 // TODO defer to Type::Instance(<bytes from typeshed>).member
                 Type::Unknown
             }
+            Type::EllipsisType => {
+                // TODO: attribute lookup on Ellipsis type
+                Type::Unknown
+            }
```
Comment (AlexWaygood) [inline crates/red_knot_python_semantic/src/types.rs:305]: I'm not sure this needs to be its own variant in the `Type` enum... it's conceptually just an instance of `types.EllipsisType`. `types.EllipsisType` was only exposed on Python 3.10+, but typeshed already provides a nice little compatibility workaround for us here: <https://github.com/python/typeshed/blob/f58dac1d62d33bb2255b762783d06463c40f5065/stdlib/builtins.pyi#L1849-L1865>

(`builtins.Ellipsis` is the same object as `...`)

```pycon
>>> ...
Ellipsis
>>> Ellipsis
Ellipsis
>>> Ellipsis is ...
True
```

File: crates/red_knot_python_semantic/src/types/infer.rs [hunk 2]
```diff
@@ -1337,8 +1337,7 @@ impl<'db> TypeInferenceBuilder<'db> {
         &mut self,
         _literal: &ast::ExprEllipsisLiteral,
     ) -> Type<'db> {
-        // TODO Ellipsis
-        Type::Unknown
+        Type::EllipsisType
```
Comment (AlexWaygood) [inline crates/red_knot_python_semantic/src/types/infer.rs:1340]: I think I would probably just revert the changes you made to `types.rs` and `types/display.rs` and simply do this here:

```suggestion
        builtins_symbol_ty_by_name(self.db, "Ellipsis")
```

It doesn't seem to get us quite the right answer currently (the typeshed stubs there are using some features we don't support right now) so you may need to adjust the assertion in your test and add a TODO. But it'll set us up better in the long run.

**Review by AlexWaygood — COMMENTED** [review]
Hmm, not sure about this approach

File: crates/red_knot_python_semantic/src/types.rs (hunk 1 above)
Comment (AlexWaygood) [reply crates/red_knot_python_semantic/src/types.rs:305]: One interesting way in which `...` _is_ special is that it's a singleton: it is the _only_ instance of `types.EllipsisType` that there is, or that there could be. But that can be thought of as just a sealed type that has exactly one member, and we've yet to implement a generalised way of handling sealed types. (We'll need them to tackle enums in Python!) See https://github.com/astral-sh/ruff/issues/12694.

File: crates/red_knot_python_semantic/src/types/infer.rs (hunk 2 above)
Comment (chriskrycho) [reply crates/red_knot_python_semantic/src/types/infer.rs:1340]: That seems reasonable. The resulting display/result is `Unknown | Literal[EllipsisType]`. As I am leaving a TODO: what *should* it be?

Comment (AlexWaygood) [reply crates/red_knot_python_semantic/src/types/infer.rs]: It should just be `types.EllipsisType` (if on py310+) or `ellipsis` (if on Python <=3.9)

File: crates/red_knot_python_semantic/src/types/infer.rs [hunk 3]
```diff
@@ -1337,8 +1337,10 @@ impl<'db> TypeInferenceBuilder<'db> {
         &mut self,
         _literal: &ast::ExprEllipsisLiteral,
     ) -> Type<'db> {
-        // TODO Ellipsis
-        Type::Unknown
+        // TODO: this does not currently resolve to `types.EllipsisType` as it
+        // should, and it needs support for sealed types (with a single member
+        // for singletons like this) as well (#12694).
```
Comment (carljm) [inline crates/red_knot_python_semantic/src/types/infer.rs:1342]: I would eliminate this TODO. This TODO will be resolved by other added features; the code right here shouldn't need to change at all, which means we also don't need a reminder TODO comment here.

```suggestion
```

**Review by carljm — APPROVED** [review]
